Add Execution Metadata Tables in MySQL and Integrate Iceberg Storage

kunwp1 opened the pull request:
This PR introduces three execution metadata tables in MySQL to reference workflow execution data stored in Iceberg tables.

### Background

Currently, workflow execution results are stored in Iceberg table format. This PR extends the same format to store:

1. **Results for each port**: Different ports produce distinct table schemas, requiring separate tables to preserve their structure.
2. **Console messages per operator**: Logs are stored at the operator level for better debugging and performance monitoring.
3. **Runtime stats for each execution**: Since runtime stats share a common schema across workflows, storing them in a single table improves efficiency.

### Changes

#### MySQL schema updates:
- `WORKFLOW_EXECUTIONS`: Added `runtime_stats_uri` column.
- `OPERATOR_EXECUTIONS`: Dropped `operator_execution_id` column and added `console_messages_uri`.
- New Table: `OPERATOR_PORT_EXECUTIONS`
```sql
CREATE TABLE IF NOT EXISTS operator_port_executions (
    workflow_execution_id INT UNSIGNED NOT NULL,
    operator_id VARCHAR(100) NOT NULL,
    port_id INT NOT NULL,
    result_uri TEXT DEFAULT NULL,
    UNIQUE (workflow_execution_id, operator_id, port_id),
    FOREIGN KEY (workflow_execution_id) REFERENCES workflow_executions (eid) ON DELETE CASCADE
);
```

#### Iceberg Table Integration:
- After creating an Iceberg document, its URI will be inserted into the corresponding MySQL table.
- The system will reference these URIs in MySQL when reading or writing to Iceberg tables.
- Assumes a properly configured MySQL environment for development.

### Migration
Run 20.sql to apply these schema changes.

<img width="378" alt="Screenshot 2025-02-04 at 3 01 23 PM" src="https://github.com/user-attachments/assets/9a0645ba-18d8-41a3-9f7f-4815720a31fd" />

kunwp1 commented:
I've added @shengquan-ni as the primary reviewer and @bobbai00 as the secondary reviewer. Thanks!

shengquan-ni commented on core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/ScheduleGenerator.scala:
```diff
@@ -186,7 +186,15 @@ abstract class ScheduleGenerator(
     DocumentFactory.createDocument(storageUri, schema)
-    ExecutionResourcesMapping.addResourceUri(workflowContext.executionId, storageUri)
+    // insert the operator port execution
+    if (!workflowContext.isTestContext) {
```
Instead of introducing a new flag in workflowContext, can you use the existing config parameter, such as UserSystemEnabled, to check? Adding a separate code flag just to differentiate between test and actual runtime is not a good practice.

kunwp1 replied:
Fixed!

shengquan-ni commented on core/amber/src/main/scala/edu/uci/ics/texera/web/service/ExecutionConsoleService.scala:
```diff
@@ -84,6 +117,21 @@ class ExecutionConsoleService(
+          )
+          writer.putOne(tuple)
+        } finally {
+          writer.close()
```
Is this suppose to be `writer.flush()`?

shengquan-ni commented on core/amber/src/main/scala/edu/uci/ics/texera/web/service/ExecutionResultService.scala:
```diff
@@ -93,14 +93,13 @@ object ExecutionResultService {
-      executionId,
-      physicalOps.head.id.logicalOpId,
-      PortIdentity()
+    val storageUri = WorkflowExecutionsResource.getResultUriByExecutionAndPort(
```
Does this require user system to be enabled?

kunwp1 replied:
Resolved!

shengquan-ni reviewed (changes requested):
I saw many places calling `WorkflowExecutionsResource.getResultUrixxxx` without checking if user system is enabled. Please add a fallback for those cases. Make sure to test it without user system.

kunwp1 commented:
I've fixed the PR and addressed the cases where the user system is not enabled. Given our discussion that developer mode always requires a MySQL database, I believe we should deprecate the user-sys configuration variable soon.

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/DocumentFactory.scala:
```diff
@@ -24,6 +29,92 @@ object DocumentFactory {
 
   private def sanitizeURIPath(uri: URI): String = uri.getPath.stripPrefix("/").replace("/", "_")
 
+  private def createIcebergDocument(
```
Since this private util func is only called once, I think there is no need to make it a function

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/DocumentFactory.scala:
```diff
@@ -24,6 +29,92 @@ object DocumentFactory {
+    )
+  }
+
+  private def loadIcebergDocument(
```
Ditto

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/FileResolver.scala:
```diff
@@ -17,6 +17,8 @@ import scala.util.{Success, Try}
 object VFSResourceType extends Enumeration {
```
in this PR can you move this VFSResourceType to `VFSURIFactory`?

bobbai00 commented on core/workflow-core/src/test/scala/edu/uci/ics/amber/storage/result/iceberg/IcebergDocumentConsoleMessagesSpec.scala:
```diff
@@ -0,0 +1,46 @@
+package edu.uci.ics.amber.storage.result.iceberg
```
For the test, you can simply mimic `IcebergDocumentSpec`: inherit `VirtualDocumentSpec` and override the generateSampleItems func

bobbai00 commented on core/gui/src/app/dashboard/type/workflow-runtime-statistics.ts:
```diff
@@ -1,11 +1,12 @@
   totalControlProcessingTime: number;
   totalIdleTime: number;
   numberOfWorkers: number;
+  status: number;
```
What does this field mean?

kunwp1 replied:
It's the state of the operator (e.g., whether it is running or completed). I just added this variable to match with the structure in the backend.

bobbai00 replied:
I see. Then some comments on this field is needed.

bobbai00 commented on core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/workflow/WorkflowExecutionsResource.scala:
```diff
@@ -96,19 +90,107 @@ object WorkflowExecutionsResource {
     }
   }
 
+  def insertResultUri(
```
rename it to `insertOperatorPortResultUri` ?

bobbai00 commented on core/amber/src/main/scala/edu/uci/ics/texera/web/service/ExecutionConsoleService.scala:
```diff
@@ -27,19 +27,52 @@ import edu.uci.ics.texera.web.model.websocket.request.python.{
 
   val bufferSize: Int = AmberConfig.operatorConsoleBufferSize
 
+  private val consoleWriters: mutable.Map[String, BufferedItemWriter[Tuple]] = mutable.Map()
```
rename it to `consoleMessageOpIdToWriterMap`

bobbai00 reviewed (commented):
Left some comments

kunwp1 commented:
Hi @shengquan-ni and @bobbai00, I've updated the PR to address all your comments. Please let me know if you have any further feedback or suggestions.

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/DocumentFactory.scala:
```diff
@@ -31,15 +27,11 @@ object DocumentFactory {
-      case _ =>
+      case DATASET_FILE_URI_SCHEME => new DatasetFileDocument(fileUri)
+      case "file"                  => new ReadonlyLocalFileDocument(fileUri)
+      case scheme =>
```
rename to `unsupportedSchema`

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/DocumentFactory.scala:
```diff
@@ -109,59 +98,54 @@ object DocumentFactory {
+              ),
+              Some(amberSchema)
+            )
+          case mode =>
```
`unsupportedMode`

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/result/IcebergTableSchema.scala:
```diff
@@ -0,0 +1,21 @@
+
+import edu.uci.ics.amber.core.tuple.{Attribute, AttributeType, Schema}
+
+object IcebergTableSchema {
```
Rename it to `ResultSchema`

bobbai00 reviewed (approved):
LGTM! Left few minors.

shengquan-ni reviewed (approved):
lgtm
